### PR TITLE
WIP: Update System.arraycopy to support null restricted value type arrays

### PR DESCRIPTION
**Pull request by a7ehuo**
Required by JEP 401: If null restricted value type is enabled, we need to preform null check on the value being stored in order to throw a `NullPointerException` if the array is null-restricted and the value to write is null. If it is this case, `System.arraycopy` cannot be transformed into arraycopy instructions.

Based on whether or not the source array and the destination array are null restricted value type, and whether or not they are flattened, VP decides whether or not transform `System.arraycopy`.

### CODE REVIEW

**Review by hzongaro — COMMENTED** [review]
Looks great!  Mostly minor comments

File: compiler/optimizer/OMRValuePropagation.cpp [hunk 1]
```diff
@@ -7355,6 +7358,33 @@ void OMR::ValuePropagation::doDelayedTransformations()
...
+            }
+
+         transformNullRestrictedArrayCopy(nullRestrictedArrayCopyTree);
+
+         if (trace())
+            {
+            TR::CFG *cfg = comp()->getFlowGraph();
+            comp()->dumpMethodTrees("Trees after transformNullRestrictedArrayCopy");
+            comp()->getDebug()->print(comp()->getOutFile(), cfg);
+            }
```
[the remark this answers is not in the document]
Comment (hzongaro) [reply compiler/optimizer/OMRValuePropagation.cpp:7384]: I noticed after making this comment that other transformations that are related to `arraycopy` also dump trees after every transformation.  I wonder whether it would be appropriate to introduce a way of requesting a more detailed dump for Value Propagation, and placing all of these individual calls to `dumpMethodTrees` under that option.  Obviously that would be a bigger change that would be for discussion, and shouldn't be done under this pull request.

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 2]
```diff
@@ -1279,6 +1287,109 @@ void OMR::ValuePropagation::transformArrayCopyCall(TR::Node *node)
...
+                  //        - (2) Both arrays are flattened, elementSize needs to be updated in order to
+                  //          use arraycopy instructions
+                  //        - (3) Either of the arrays might or might be flattened, System.arraycopy
+                  //          should not be transformed into arraycopy instructions.
+                  //
+                  if (isValueTypeArrayFlatteningEnabled)
+                     {
+                     if (trace())
+                        traceMsg(comp(), "%s: n%dn %p  isArrayElementFlattened dst %d src %d\n", __FUNCTION__, node->getGlobalIndex(), node,
+                           isArrayElementFlattened(dstObject), isArrayElementFlattened(srcObject));
```
[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/ValuePropagationCommon.cpp:1343]: >you could just open an issue and mark it as "Help wanted" so that the various places that print values of that type could be cleaned up all at once

This is a good idea! I'll create an issue for this work.

File: compiler/optimizer/OMRValuePropagation.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/OMRValuePropagation.cpp:7384]: > it would be sufficient to dump the trees just before the first arraycopy transformation and after the last

Because there are other sub arraycopy transformations happen before `transformNullRestrictedArrayCopy`, the trees have changed quite a bit compared to the initial trees before the first arraycopy transformation. Dumping before and after trees here is useful to understand what `transformNullRestrictedArrayCopy` does here.

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/ValuePropagationCommon.cpp:1343]: Created https://github.com/eclipse-openj9/openj9/issues/17956

File: compiler/optimizer/OMRValuePropagation.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/OMRValuePropagation.cpp:7384]: https://github.com/eclipse-openj9/openj9/issues/17957 is created

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 3]
```diff
@@ -3310,6 +3545,222 @@ void OMR::ValuePropagation::transformRTMultiLeafArrayCopy(TR_RealTimeArrayCopy *
...
+
+      tmpBlock = tmpBlock->getPrevBlock();
+      TR::TreeTop *bbStart = tmpBlock->getEntry();
+      TR::TreeTop *bbEnd = tmpBlock->getExit();
+
+      if (bbStart->getNextTreeTop() != bbEnd)
+         {
+         isPrevBlockOfExtendedBlockEmpty = false;
+         break;
+         }
```
[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/ValuePropagationCommon.cpp:3673]: > Did you want isPrevBlockOfExtendedBlockEmpty to be false if there is at least one preceding block in the sequence of blocks that is not empty (which is what is happening now) or did you want it to be false only if the first block in the sequence is not empty?

`isPrevBlockOfExtendedBlockEmpty` should be `false` as long as there is at least one preceding block is not empty. Any IL trees in those preceding blocks are not the original next IL trees following the original `System.arraycopy` call. 

Using the above example,  `isPrevBlockOfExtendedBlockEmpty` is `false`. We need to split the block at `block_13`. 

As [per comment above](https://github.com/eclipse/omr/pull/7084#discussion_r1291518424), I'm adding comments to this part of the code. 
```
n99n BBStart <block_11>
n98n BBEnd <block_11>
n97n BBStart <block_12> (extension of previous block)
n96n istore ...
 ...
n95n BBEnd <block_12> (extension of previous block)
n94n BBStart <block_13> (extension of previous block) //<--- next block containing the original next IL trees
   ...
```
==>
```
n99n BBStart <block_11>
n98n BBEnd <block_11>
n97n BBStart <block_12> (extension of previous block)
n96n istore ...
 ...
n95n BBEnd <block_12>

n94n BBStart <block_13> (extension of previous block) //<--- slow block that has the call should jump to here
   ...
```

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 4]
```diff
@@ -1289,6 +1400,123 @@ void OMR::ValuePropagation::transformArrayCopyCall(TR::Node *node)
...
+               n55n          iload  <temp slot 6>[#432  Auto]
+               n57n          iload  <temp slot 7>[#433  Auto]
+               n40n      BBEnd </block_-1>
+            */
+
+         // Create the block that contains the System.arraycopy call which will be the slow path
+         TR::Block *slowBlock = TR::Block::createEmptyBlock(node, comp(), UNKNOWN_COLD_BLOCK_COUNT);
+         slowBlock->setIsCold(true);
+
+         TR::TreeTop *newCallTree = _curTree->duplicateTree();
```
[the remark this answers is not in the document]
Comment (hzongaro) [reply compiler/optimizer/ValuePropagationCommon.cpp:1457]: > Could you elaborate a bit more on "all the children of the duplicated call are going to be discarded"?

I just meant that `TR::TreeTop::duplicateTree` will duplicate every node in the call tree, but a little later every child of the duplicate `System.arraycopy` call will be [replaced with a load of a temporary](https://github.com/eclipse/omr/blob/8bb98cf3e067a33212f6613eca7b47991a7b8869/compiler/optimizer/ValuePropagationCommon.cpp#L1484-L1490).  It seems unnecessary to duplicate those five subtrees when they will be immediately replaced.

[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/ValuePropagationCommon.cpp:1457]: Thank for the clarification! Now I see what is referred here. 

>but a little later every child of the duplicate System.arraycopy call will be [replaced with a load of a temporary](https://github.com/eclipse/omr/blob/8bb98cf3e067a33212f6613eca7b47991a7b8869/compiler/optimizer/ValuePropagationCommon.cpp#L1484-L1490).

Here the invoked call is`OMR::TreeTop::duplicateTree()`, not `OMR::Node::duplicateTree(bool duplicateChildren = true)`. The duplicated treetop node `n41n` has `n42n` as the first child. `n42n` is not replaced. It is the children of `n42n` that are replaced later, not the children of `n41n`. 

I guess I can call `OMR::Node::duplicateTree(bool duplicateChildren)` with `duplicateChildren` as `false` and create a treetop for the new call node.

**Comment by a7ehuo** [comment]
@hzongaro All comments are addressed in d0bf0ce. Ready for another review. Thanks!

**Comment by a7ehuo** [comment]
x86-64 macOS test failed. It looks to be an known issue https://github.com/eclipse/omr/issues/6516

```
2023-08-17T19:42:57.2705020Z 32: [  FAILED  ] 2 tests, listed below:
2023-08-17T19:42:57.2705960Z 32: [  FAILED  ] PortSockTest.poll_functionality_basic
2023-08-17T19:42:57.2706840Z 32: [  FAILED  ] PortSockTest.poll_functionality_many_sockets
2023-08-17T19:42:57.2707870Z 32: 
2023-08-17T19:42:57.2708790Z 32:  2 FAILED TESTS
2023-08-17T19:42:57.2709630Z 32:   YOU HAVE 3 DISABLED TESTS
```

```
The following tests FAILED:
	 32 - porttest (Failed)
```

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 5]
```diff
@@ -1289,6 +1428,136 @@ void OMR::ValuePropagation::transformArrayCopyCall(TR::Node *node)
...
+
+         TR::Node *oldCallNode = node;
+         if (trace())
+            traceMsg(comp(),"Creating temps for children of the original call node n%dn %p. new call node n%dn %p\n", oldCallNode->getGlobalIndex(), oldCallNode, newCallNode->getGlobalIndex(), newCallNode);
+
+         // Create temporaries for System.arraycopy arguments and replace the children of the new call node with the temps
+         for (int32_t i = 0 ; i < oldCallNode->getNumChildren(); ++i)
+            {
+            TR::Node *child = oldCallNode->getChild(i);
+            TR::SymbolReference *newSymbolReference = comp()->getSymRefTab()->createTemporary(comp()->getMethodSymbol(), child->getDataType());
```
Comment (hzongaro) [inline compiler/optimizer/ValuePropagationCommon.cpp:1504]: I think the temporary is only used if `isLoadConst()` is false, so I think this ought to be moved under the `else` below.

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 6]
```diff
@@ -1289,6 +1428,136 @@ void OMR::ValuePropagation::transformArrayCopyCall(TR::Node *node)
...
+               n4n           iconst 0
+               n5n           aload  <parm 1 [LSomeInterface;>[#420  Parm]
+               n6n           iconst 0
+               n7n           iload  TestSystemArraycopy4.ARRAY_SIZE I[#421  Static]
+
+               ==== After ===
+               n48n      astore  <temp slot 3>[#429  Auto]
+               n3n         aload  <parm 0 [LSomeInterface;>[#419  Parm]
+               n50n      istore  <temp slot 4>[#430  Auto]
+               n4n         iconst 0
```
Comment (hzongaro) [inline compiler/optimizer/ValuePropagationCommon.cpp:1453]: I think with your recent change to only use temporaries for non-constant values, the stores of the constants should be removed from these comments.  Similarly, the commoning arrows shown below for `n4n` and `n6n` need to be removed, and the loads of `#430` and `#432` in the slow block should become `iconst 0` nodes.

**Review by hzongaro — COMMENTED** [review]
A few minor comments, plus I think there might be another case where a run-time test for null-restricted component types is needed.

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 7]
```diff
@@ -1279,6 +1288,136 @@ void OMR::ValuePropagation::transformArrayCopyCall(TR::Node *node)
...
+               }
+            case TR_maybe:
+               {
+               if (isSrcArrayCompTypePrimitiveValueType != TR_no)
+                  {
+                  // The source might or might not be a primitive VT array and the target might or might not
+                  // be a primitive VT array, add runtime check before the arraycopy instructions
+                  doRuntimeNullRestrictedTest = true;
+                  }
+               // else: The source is not primitive VT array and the target might or might not a primitive VT,
```
[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/ValuePropagationCommon.cpp:1379]: >"else: The source is a primitive VT array" ...

In the "else" case, `isSrcArrayCompTypePrimitiveValueType` == `TR_no`, meaning the source is not primitive VT array. I'm not sure why it should be `"else: The source is a primitive VT array"`

File: compiler/optimizer/ValuePropagationCommon.cpp [hunk 8]
```diff
@@ -1279,6 +1288,136 @@ void OMR::ValuePropagation::transformArrayCopyCall(TR::Node *node)
...
+               //       nothing needs to be done here because:
+               //       - If in runtime the destination is primitive VT array, it should fail array store check
```
[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/ValuePropagationCommon.cpp:1381]: I'll run experiment to double check this case

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (hzongaro) [reply compiler/optimizer/ValuePropagationCommon.cpp:1379]: Sorry!  I was getting myself confused.  I'll mark this as resolved.

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/ValuePropagationCommon.cpp:1381]: > What if the source array is a fixed array of type java/lang/Object and the target turns out to be a flattened primitive value type array at runtime? 

I reproduced issue with this case. Good catch! I think if destination array might or might not be the primitive vt array, we need to add runtime check regardless of the source array type. 

```c++
            case TR_maybe:
               {
               if (isSrcArrayCompTypePrimitiveValueType != TR_no)
                  {
                  // isSrcArrayCompTypePrimitiveValueType either equals to TR_yes or TR_maybe 
                  // The source might or might not be a primitive VT array and the target might or might not
                  // be a primitive VT array, add runtime check before the arraycopy instructions
                  doRuntimeNullRestrictedTest = true;
                  }
               // else: The source is not primitive VT array and the target might or might not a primitive VT,
               //       nothing needs to be done here because:
               //       - If in runtime the destination is primitive VT array, it should fail array store check
               //       - If in runtime the destination is not primitive VT array, it should function as it is today
               break;
               }
```
==>
```c++
            case TR_maybe:
               {
               doRuntimeNullRestrictedTest = true;
               break;
               }
```

[the remark this answers is not in the document]
Comment (a7ehuo) [reply compiler/optimizer/ValuePropagationCommon.cpp]: Adressed in 71ddb70

File: compiler/optimizer/ValuePropagationCommon.cpp (hunk 5 above)
Comment (a7ehuo) [reply compiler/optimizer/ValuePropagationCommon.cpp]: Adressed in 71ddb70

**Comment by a7ehuo** [comment]
@hzongaro All comments are addressed. Ready for another review. Thanks!

**Review by hzongaro — APPROVED** [review]
Looks good.  Thanks!

**Comment by hzongaro** [comment]
Jenkins build all
